src/create-template-orders.js

```diff
@@ -20,20 +33,263 @@ async function _fetchLastStartTimestamp() {
+  const cronExpression = parser.parseExpression(schedule)
+  let nextDeliveryCronDate = cronExpression.next()
+  const nextDeliveryDate = nextDeliveryCronDate.toDate()
+  let nextDeliveryDateISOString = nextDeliveryDate.toISOString()
+  const cutoffDays = lineItem.custom.fields.cutoffDays
+  if (cutoffDays) {
+    nextDeliveryDate.setDate(nextDeliveryDate.getDate() - cutoffDays)
+    if (nextDeliveryDate.toISOString() < checkoutOrder.createdAt) {
+      nextDeliveryDateISOString = nextDeliveryDate.toISOString()
+      nextDeliveryCronDate = cronExpression.next()
+    }
+  }
+  const reminderDays = lineItem.custom.fields.reminderDays
+  let nextReminderDateISOString
+  if (reminderDays) {
+    nextDeliveryCronDate.setDate(nextDeliveryCronDate.getDate() - reminderDays)
+    nextReminderDateISOString = nextDeliveryCronDate.toISOString()
+  }
+  const templateOrder = {
+    orderNumber:
+      lineItem.custom.fields.subscriptionKey +
+      (quantityIncrement ? `-${quantityIncrement}` : ''),
+    customerEmail: checkoutOrder.customerEmail,
+    totalPrice: lineItem.totalPrice,
+    lineItems: [lineItem],
+    custom: {
+      type: {
+        typeId: 'type',
+        key: 'subscription-template-order',
+      },
+      fields: {
+        ...checkoutOrder.custom.fields,
+        nextDeliveryDate: nextDeliveryDateISOString,
+        schedule,
+        checkoutOrderRef: { typeId: 'order', id: checkoutOrder.id },
+        reminderDays,
+        nextReminderDate: nextReminderDateISOString,
+      },
+    },
+    inventoryMode: 'None',
+    state: {
+      typeId: 'state',
+      key: 'Active',
+    },
+  }
+  delete templateOrder.custom.fields.hasSubscription
+  return templateOrder
+}
+
+function _createPaymentDraft(payment) {
+  const paymentDraft = _.cloneDeep(payment)
+  delete paymentDraft.key
+  return paymentDraft
+}
+
+async function _buildFetchCheckoutOrdersUri() {
   let where =
     'custom(fields(hasSubscription=true)) AND custom(fields(isSubscriptionProcessed is not defined))'
   const lastStartTimestamp = await _fetchLastStartTimestamp()
   if (lastStartTimestamp)
-    where = `createdAt > ${lastStartTimestamp} AND ${where}`
-  const {
-    body: { results },
-  } = await apiRoot.orders().get({ queryArgs: { where } }).execute()
-  return results
+    where = `createdAt > "${lastStartTimestamp.value}" AND ${where}`
+
+  const uri = ctpClient.builder.orders
+    .where(where)
+    .expand('paymentInfo.payments[*]')
+  return uri
 }
 
-async function createTemplateOrders() {
-  const unprocessedCheckoutOrders = await fetchCheckoutOrders()
+async function _updateLastStartTimestamp(startDate) {
+  await apiRoot
+    .customObjects()
+    .post({
+      body: {
+        container: LAST_START_TIMESTAMP_CUSTOM_OBJECT_CONTAINER,
+        key: LAST_START_TIMESTAMP_CUSTOM_OBJECT_KEY,
+        value: startDate.toISOString(),
+      },
+    })
+    .execute()
+}
+
+async function createTemplateOrders(startDate) {
```

Comment:
I think this function should also provide some metrics for observability to be logged, for instance how many template orders are created, failed, skipped, processed etc. Wdyt ?

Reply:
https://github.com/commercetools/commercetools-subscriptions/pull/5/commits/25dce97a8a40d5a0278f3e7e3d96de8837c64546